### PR TITLE
Harden blueprint update and metadata handling and refactor test suite

### PR DESCRIPTION
**Pull request by luuquangvu**
## Summary by Sourcery

Harden blueprint update, metadata handling, and coordinator/networking logic while refactoring and expanding the test suite structure.

New Features:
- Add comprehensive tests for coordinator update cycles, storage/metadata handling, networking, and edge cases, including backup/restore flows and Git diff caching.

Bug Fixes:
- Prevent crashes and stale state when processing invalid YAML or unexpected exceptions during blueprint content processing and risk detection.
- Ensure unsafe paths and URLs are rejected for blueprint installation and updates, invalidating cached metadata where appropriate.
- Fix domain inference for blueprints scanned from nested folder structures and align automation/script reload behavior with normalized domains.
- Avoid ghost or spurious updates by normalizing YAML content and comparing canonical hashes, ignoring comment-only changes.
- Prevent KeyError and stale ETag/remote-hash reuse across cold starts and after source_url changes, including ETag invalidation on hash mismatches.
- Harden handling of malformed persisted metadata, invalid config values, networking failures, and HTTP redirect/security edge cases.

Enhancements:
- Rework source_url injection to always use structured YAML parsing and canonical dumping, removing brittle regex-based manipulation and supporting non-string inputs more safely.
- Improve blueprint scanning, metadata reuse, and pruning so persisted hashes/ETags are rehydrated only on first update and stale entries are cleaned up.
- Tighten auto-update and notification flows, including multi-domain reloads, concurrency limiting for background refresh, and proper task cancellation/shutdown behavior.
- Refine utilities for filter mode and selected blueprint normalization, plus configuration-driven CDN usage for fetches.
- Improve error reporting via structured last_error codes, update_blocking_reason propagation, and clearer logging around processing and networking failures.

Build:
- Update pytest configuration to include pytest-timeout and enforce a 90% coverage threshold with a global per-test timeout.

Documentation:
- Simplify README and localized documentation by removing FAQ sections, updating terminology, and clarifying source_url requirements and risk report labels.

Tests:
- Restructure tests into coordinator, entities, ui, and utils packages and significantly expand coverage for coordinator core, networking, storage, metadata, edge cases, entities, and UI/config flows.
- Add new safety and normalization tests for path/URL validation, filter/selection hardening, YAML/source_url handling, backup rotation, and retry logic.
- Introduce higher coverage threshold enforcement and test timeout configuration in pytest settings.

## Summary by CodeRabbit

* **Documentation**
  * Removed FAQ (EN/VI); updated Advanced Compatibility Guard wording and example blueprint title.

* **Bug Fixes & Improvements**
  * Toughened URL/redirect safety, centralized error sanitization, improved metadata/hash handling, and more robust update/install flows.

* **UI Updates**
  * Breaking-risks heading relabeled to warning-style "POTENTIAL BREAKING CHANGES" and messaging updated across locales.

* **Tests**
  * Large expansion of coordinator, networking, storage, metadata, update-cycle and UI tests; some legacy tests removed.

* **Chores**
  * Added pre-commit/pytest config and tightened an integration dependency constraint.

### CODE REVIEW

**Comment by luuquangvu** [comment]
@coderabbitai full review

File: custom_components/blueprints_updater/coordinator.py [hunk 1]
```diff
@@ -2773,61 +2777,49 @@ def _hash_content(content: str, already_normalized: bool = False) -> str:
...
-                return yaml_util.dump(parsed)
-            except Exception as err:
-                _LOGGER.warning("Structured YAML injection failed for %s: %s", source_url, err)
-
-        return content
+        try:
+            return yaml_util.dump(parsed)
+        except (yaml.YAMLError, TypeError, ValueError) as err:
+            _LOGGER.warning("YAML canonicalization failed for %s: %s", source_url, err)
+            return BlueprintUpdateCoordinator._normalize_content(content)
```
[the remark this answers is not in the document]
Comment (luuquangvu) [reply custom_components/blueprints_updater/coordinator.py:2902]: This PR fully resolves issue #45 by treating the local `source_url` as the single source of truth.

File: README.md [hunk 2]
```diff
@@ -156,35 +153,13 @@ To apply changes instantly without waiting for the next scheduled background sca
...
 ## Requirements
 
 For a blueprint to be trackable, it **must** contain a valid `source_url` within its metadata:
 
 ```yaml
 blueprint:
-  name: "My Blueprint"
+  name: "Blueprint Name"
   source_url: https://github.com/user/repo/blob/main/blueprint.yaml
   # ...
```
[the remark this answers is not in the document]
Comment (luuquangvu) [reply README.md:164]: This PR fully resolves issue #45 by treating the local `source_url` as the single source of truth. Home Assistant retains the URL entered by the user when importing a blueprint.

File: custom_components/blueprints_updater/coordinator.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (luuquangvu) [reply custom_components/blueprints_updater/coordinator.py:2902]: The `_ensure_source_url` function also checks if the content is invalid; if it does, it will not inject the `source_url`.

**Comment by luuquangvu** [comment]
@coderabbitai full review

**Comment by luuquangvu** [comment]
@coderabbitai full review

**Comment by luuquangvu** [comment]
@coderabbitai pause

**Comment by luuquangvu** [comment]
@coderabbitai review

**Comment by luuquangvu** [comment]
@coderabbitai Our current mechanism still ensures absolute security because the final destination must be HTTPS. If an HTTP link does not redirect to HTTPS, it will still be blocked at the last step.

**Comment by luuquangvu** [comment]
@coderabbitai Since these are blueprints with public URLs, not sensitive data, blocking HTTP from the outset is unnecessary. This could lead to situations where blueprints with `source_url` calling GitHub via HTTP links are blocked and cannot be updated.

**Comment by luuquangvu** [comment]
@coderabbitai review

**Comment by luuquangvu** [comment]
@coderabbitai According to official documents, use [tool.pytest] to leverage native TOML types (supported since pytest 9.0)

**Comment by luuquangvu** [comment]
@coderabbitai the project already includes a specific Prettier version in `package.json` and `package-lock.json`. Both `ty` and `pyright` are required for use of the type checker effectively.

File: .pre-commit-config.yaml [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+        language: system
+        types: [python]
+        pass_filenames: false
+        stages: [pre-push]
+
+      - id: prettier
+        name: prettier
+        entry: npx prettier --write
+        language: system
+        types_or: [json, yaml, markdown, javascript, toml]
```
[the remark this answers is not in the document]
Comment (luuquangvu) [reply .pre-commit-config.yaml]: @coderabbitai `prettier-plugin-toml` is also already included in `package.json`

**Comment by luuquangvu** [comment]
@coderabbitai review

**Comment by luuquangvu** [comment]
@coderabbitai review

**Comment by luuquangvu** [comment]
@coderabbitai review
